File: rust/src/bin/http_rest.rs
use axum::{
    Json, Router,
    extract::{Path, State},
    http::StatusCode,
    routing::get,
};
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, sync::Arc};
use tokio::sync::Mutex;

#[derive(Clone, Serialize)]
struct User {
    id: u64,
    name: String,
}

#[derive(Deserialize)]
struct NewUser {
    name: String,
}

type Store = Arc<Mutex<HashMap<u64, User>>>;

async fn all_users(State(store): State<Store>) -> Json<Vec<User>> {
    Json(store.lock().await.values().cloned().collect())
}

async fn get_user(
    State(store): State<Store>,
    Path(id): Path<u64>,
) -> Result<Json<User>, StatusCode> {
    store.lock().await.get(&id).cloned().map(Json).ok_or(StatusCode::NOT_FOUND)
}

async fn create_user(
    State(store): State<Store>,
    Json(input): Json<NewUser>,
) -> (StatusCode, Json<User>) {
    let mut users = store.lock().await;
    let id = users.len() as u64 + 1;
    let user = User {
        id,
        name: input.name,
    };
    users.insert(id, user.clone());
    (StatusCode::CREATED, Json(user))
}

#[tokio::main]
async fn main() {
    let store = Arc::new(Mutex::new(HashMap::new()));
    let app = Router::new()
        .route("/users", get(all_users).post(create_user))
        .route("/users/{id}", get(get_user))
        .with_state(store);

    let listener = tokio::net::TcpListener::bind("127.0.0.1:3000").await.unwrap();
    axum::serve(listener, app).await.unwrap();
}

File: python/src/http_rest.py
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel


class User(BaseModel):
    id: int
    name: str


class NewUser(BaseModel):
    name: str


app = FastAPI()
users: dict[int, User] = {}


@app.get("/users")
def list_users() -> list[User]:
    return list(users.values())


@app.get("/users/{user_id}")
def get_user(user_id: int) -> User:
    if user := users.get(user_id):
        return user
    raise HTTPException(404)


@app.post("/users", status_code=201)
def create_user(body: NewUser) -> User:
    user = User(id=len(users) + 1, name=body.name)
    users[user.id] = user
    return user


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="127.0.0.1", port=3000)
